src/hotspot/share/prims/jvmtiTagMapTable.hpp

```diff
@@ -25,67 +25,73 @@
 #ifndef SHARE_VM_PRIMS_TAGMAPTABLE_HPP
 #define SHARE_VM_PRIMS_TAGMAPTABLE_HPP
 
+#include "gc/shared/collectedHeap.hpp"
+#include "memory/allocation.hpp"
 #include "oops/weakHandle.hpp"
 #include "utilities/hashtable.hpp"
+#include "utilities/resizeableResourceHash.hpp"
 
 class JvmtiEnv;
-
-// Hashtable to record oops used for JvmtiTagMap
 class JvmtiTagMapEntryClosure;
 
-class JvmtiTagMapEntry : public HashtableEntry<WeakHandle, mtServiceability> {
-  jlong _tag;                           // the tag
- public:
-  JvmtiTagMapEntry* next() const {
-    return (JvmtiTagMapEntry*)HashtableEntry<WeakHandle, mtServiceability>::next();
-  }
```

Comment:
This should add a comment above JvmtiTagMapEntry that the entry is the key (could be renamed JvmtiTagMapKey).  Also add a comment above that the oop is needed for lookup rather than creating a WeakHandle during lookup because the HeapWalker may walk soon to be dead objects and creating a WeakHandle for an otherwise dead object makes G1 unhappy.